syntax = "proto3";

// Cross platform definitions
option java_multiple_files = true;
option java_package = "by.heshu.likelib.rpc";
option java_outer_classname = "LikelibNodeProto";
option objc_class_prefix = "LKL";

// Service and messages definitions
package likelib;

service NodePublicInterface {

    rpc get_account (Address) returns (AccountInfo) {
    }

    rpc get_node_info (None) returns (NodeInfo) {
    }

    rpc get_block (Hash) returns (Block) {
    }

    rpc get_transaction (Hash) returns (Transaction) {
    }

    rpc push_transaction (Transaction) returns (TransactionStatus) {
    }

    rpc get_transaction_result (Hash) returns (TransactionStatus) {
    }

}

//=====================================


message NodeInfo {
    Hash top_block_hash = 1;
    uint32 interface_version = 2;
    uint32 peers_number = 3;
}


message Block {
    Address coinbase = 1;
    Time timestamp = 2;
    uint64 depth = 3;
    uint64 nonce = 4;
    Hash previous_block_hash = 5;
    repeated Transaction transactions = 6;
}


message AccountInfo {
    Address address = 1;
    CurrencyAmount balance = 2;
    uint64 nonce = 3;
    repeated Hash hashes = 4;
}


message Transaction {
    Address from = 1;
    Address to = 2;
    CurrencyAmount value = 3;
    CurrencyAmount fee = 4;
    Time creation_time = 5;
    Data data = 6;
    Signature signature = 7;
}


message TransactionStatus {
    enum StatusCode {
        Success = 0;
        Rejected = 1;
        Revert = 2;
        Failed = 3;
    }
    StatusCode status = 1;
    string message = 2;
    CurrencyAmount gas_left = 3;
}


message Signature {
    string signature_bytes_at_base_64 = 1;
}


message Time {
    uint64 since_epoch = 1;
}


message CurrencyAmount {
    uint64 value = 1;
}


message Address {
    string address_at_base_58 = 1;
}


message Hash {
    string bytes_base_64 = 1;
}


message Data {
    string bytes_base_64 = 1;
}


message None {
}
